### PR TITLE
Implementing Prov custom payload

### PR DESCRIPTION
**Pull request by jebrando**
# Checklist
- [ ] I have read the [contribution guidelines] (https://github.com/Azure/azure-iot-sdk-c/blob/master/.github/CONTRIBUTING.md).
- [ ] I added or modified the existing tests to cover the change (we do not allow our test coverage to go down).
- If this is a modification that impacts the behavior of a public API
  - [ ] I edited the corresponding document in the `devdoc` folder and added or modified requirements.
- I submitted this PR against the correct branch: 
  - [ ] This pull-request is submitted against the `master` branch. 
  - [ ] I have merged the latest `master` branch prior to submission and re-merged as needed after I took any feedback.
  - [ ] I have squashed my changes into one with a clear description of the change.

# Reference/Link to the issue solved with this PR (if any)

# Description of the problem

# Description of the solution

### CODE REVIEW

File: provisioning_client/inc/azure_prov_client/internal/prov_transport_utils.h [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+#define PROV_TRANSPORT_UTIL_H
+
+#ifdef __cplusplus
+extern "C" {
+#include <cstdint>
+#else
+#include <stdint.h>
+#endif /* __cplusplus */
+
+uint32_t parse_retry_after_value(const char* retry_after);
```
Comment (massand) [inline provisioning_client/inc/azure_prov_client/internal/prov_transport_utils.h]: We don't need a new header for this. We could accomplish the same in prov_transport.h

**Comment by massand** [comment]
e2e tests are missing from here. Please add them

File: provisioning_client/src/prov_device_ll_client.c [hunk 2]
```diff
@@ -204,7 +214,11 @@ static void on_transport_error(PROV_DEVICE_TRANSPORT_ERROR transport_error, void
 static PROV_DEVICE_TRANSPORT_STATUS retrieve_status_type(const char* prov_status)
 {
     PROV_DEVICE_TRANSPORT_STATUS result;
-    if (prov_status == NULL || strcmp(prov_status, PROV_UNASSIGNED_STATUS) == 0)
+    if (prov_status == NULL)
+    {
+        result = PROV_DEVICE_TRANSPORT_STATUS_ERROR;
```
Comment (massand) [inline provisioning_client/src/prov_device_ll_client.c]: >PROV_DEVICE_TRANSPORT_STATUS_ERROR [](start = 17, length = 34)

In a normal world, we would do a null check right after we obtained the JSON in the calling function. This isn't a 'request failed' scenario. So, I would treat this as a local issue and failfast in the calling function and remove this NULL check here.

File: provisioning_client/src/prov_device_ll_client.c [hunk 3]
```diff
@@ -251,13 +265,26 @@ static int retrieve_json_number(JSON_Object* json_object, const char* field_name
     return result;
 }
 
-static char* retrieve_json_item(JSON_Object* json_object, const char* field_name)
+static void retrieve_json_payload(JSON_Object* json_object, PROV_INSTANCE_INFO* prov_info)
+{
+    JSON_Value* json_field;
+    // Returned Data is not available
```
Comment (massand) [inline provisioning_client/src/prov_device_ll_client.c]: >// [](start = 4, length = 2)

nit style: empty line before comment

File: provisioning_client/src/prov_transport_amqp_common.c [hunk 4]
```diff
@@ -245,6 +277,9 @@ static void on_amqp_send_complete(void* user_ctx, MESSAGE_SEND_RESULT send_resul
         PROV_TRANSPORT_AMQP_INFO* amqp_info = (PROV_TRANSPORT_AMQP_INFO*)user_ctx;
         if (send_result != MESSAGE_SEND_OK)
         {
+            (void)delivery_state;
+            // To be included when a issue in the service is fixed.
```
Comment (massand) [inline provisioning_client/src/prov_transport_amqp_common.c]: >// To be included when a issue in the service is fixed. [](start = 12, length = 55)

Is this TODO fixed?

File: provisioning_client/tests/prov_transport_http_client_ut/prov_transport_http_client_ut.c [hunk 5]
```diff
@@ -1157,7 +1177,7 @@ BEGIN_TEST_SUITE(prov_transport_http_client_ut)
         //arrange
         PROV_DEVICE_TRANSPORT_HANDLE handle = prov_dev_http_transport_create(TEST_URI_VALUE, TRANSPORT_HSM_TYPE_TPM, TEST_SCOPE_ID_VALUE, TEST_DPS_API_VALUE, on_transport_error, NULL);
```
Comment (massand) [inline provisioning_client/tests/prov_transport_http_client_ut/prov_transport_http_client_ut.c]: >TRANSPORT_HSM_TYPE_TPM [](start = 93, length = 22)

There doesn't appear to be any happy path testing for PROV_TRANSPORT_HTTP_INFO when it has an ek value. We should have a TEST_FUNCTION for that happy path because it generates a json payload specifically with those values and needs to be cleaned up.

File: provisioning_client/tests/prov_transport_mqtt_common_ut/prov_transport_mqtt_common_ut.c [hunk 6]
```diff
@@ -1344,7 +1373,51 @@ BEGIN_TEST_SUITE(prov_transport_mqtt_common_ut)
...
+        prov_transport_common_mqtt_dowork(handle);
+        g_operation_cb(TEST_MQTT_CLIENT_HANDLE, MQTT_CLIENT_ON_CONNACK, &connack, g_msg_recv_callback_context);
+        prov_transport_common_mqtt_dowork(handle);
+        g_operation_cb(TEST_MQTT_CLIENT_HANDLE, MQTT_CLIENT_ON_SUBSCRIBE_ACK, &suback, g_msg_recv_callback_context);
+        prov_transport_common_mqtt_dowork(handle);
+        g_target_transport_status = PROV_DEVICE_TRANSPORT_STATUS_ASSIGNING;
+        umock_c_reset_all_calls();
+
+        //arrange
+        STRICT_EXPECTED_CALL(mqttmessage_getTopicName(IGNORED_PTR_ARG)).SetReturn("$dps/registrations/res/202/?$rid=1&retry-after=8");
```
Comment (massand) [inline provisioning_client/tests/prov_transport_mqtt_common_ut/prov_transport_mqtt_common_ut.c]: >$dps/registrations/res/202/?$rid=1&retry-after=8 [](start = 83, length = 48)

nit: there's a handful of magic values, such as these topic values, lying around in this file. We should move them up top to maintain consistent style between UTs.

File: provisioning_client/tests/prov_transport_mqtt_ws_client_ut/prov_transport_mqtt_ws_client_ut.c [hunk 7]
```diff
@@ -378,10 +380,10 @@ BEGIN_TEST_SUITE(prov_transport_mqtt_ws_client_ut)
     TEST_FUNCTION(prov_transport_mqtt_ws_register_device_succeed)
     {
         //arrange
-        STRICT_EXPECTED_CALL(prov_transport_common_mqtt_register_device(TEST_DPS_HANDLE, on_transport_json_parse, NULL));
+        STRICT_EXPECTED_CALL(prov_transport_common_mqtt_register_device(TEST_DPS_HANDLE, on_transport_json_parse, on_transport_create_json_payload, NULL));
 
         //act
-        int result = prov_mqtt_ws_transport_register_device(TEST_DPS_HANDLE, on_transport_json_parse, NULL);
+        int result = prov_mqtt_ws_transport_register_device(TEST_DPS_HANDLE, on_transport_json_parse, on_transport_create_json_payload, NULL);
```
Comment (massand) [inline provisioning_client/tests/prov_transport_mqtt_ws_client_ut/prov_transport_mqtt_ws_client_ut.c]: >on_transport_create_json_payload [](start = 102, length = 32)

A suggestion (which might have some trivial solutions, might not): It appears we are shimming the json payload creation process in every test in this PR. We should have atleast one test that covers obtaining fake ek_value and validating the payload generated by actual production code is passed through to DoWork.

**Comment by jebrando** [comment]
We need a new service client to have e2e tests.  Should we invest in a service client or not?

---
In reply to: [488879595](https://github.com/Azure/azure-iot-sdk-c/pull/984#issuecomment-488879595) [](ancestors = 488879595)

File: provisioning_client/inc/azure_prov_client/internal/prov_transport_utils.h (hunk 1 above)
Comment (jebrando) [reply provisioning_client/inc/azure_prov_client/internal/prov_transport_utils.h]: The source will need to live somewhere.  prov_transport.h doesn't have a matching source file.

---
In reply to: [280640096](https://github.com/Azure/azure-iot-sdk-c/pull/984#discussion_r280640096) [](ancestors = 280640096)

File: provisioning_client/src/prov_transport_amqp_common.c (hunk 4 above)
Comment (jebrando) [reply provisioning_client/src/prov_transport_amqp_common.c]: It's a big part of the feature, but not everything.  They have a bug and they don't want to deploy till after build.

---
In reply to: [281245051](https://github.com/Azure/azure-iot-sdk-c/pull/984#discussion_r281245051) [](ancestors = 281245051)

File: provisioning_client/tests/prov_transport_mqtt_ws_client_ut/prov_transport_mqtt_ws_client_ut.c (hunk 7 above)
Comment (jebrando) [reply provisioning_client/tests/prov_transport_mqtt_ws_client_ut/prov_transport_mqtt_ws_client_ut.c]: MQTT doesn't support TPM.  I'll add that to http and amqp.

---
In reply to: [281253174](https://github.com/Azure/azure-iot-sdk-c/pull/984#discussion_r281253174) [](ancestors = 281253174)

File: provisioning_client/tests/prov_transport_http_client_ut/prov_transport_http_client_ut.c (hunk 5 above)
Comment (jebrando) [reply provisioning_client/tests/prov_transport_http_client_ut/prov_transport_http_client_ut.c]: The test is on the http_dowork_success.  It does TPM with ek and srk

---
In reply to: [281249536](https://github.com/Azure/azure-iot-sdk-c/pull/984#discussion_r281249536) [](ancestors = 281249536)

File: provisioning_client/inc/azure_prov_client/internal/prov_transport_utils.h (hunk 1 above)
Comment (jebrando) [reply provisioning_client/inc/azure_prov_client/internal/prov_transport_utils.h]: Moving the code to an h only situation causes unintended consequences.  The Unittests have issues with doubly defined symbols.

---
In reply to: [281403483](https://github.com/Azure/azure-iot-sdk-c/pull/984#discussion_r281403483) [](ancestors = 281403483,280640096)

**Comment by jebrando** [comment]
Working on the e2e with our new schemes

---
In reply to: [489824105](https://github.com/Azure/azure-iot-sdk-c/pull/984#issuecomment-489824105) [](ancestors = 489824105,488879595)

File: provisioning_client/inc/azure_prov_client/internal/prov_transport_utils.h (hunk 1 above)
Comment (massand) [reply provisioning_client/inc/azure_prov_client/internal/prov_transport_utils.h]: static inline should work, the compiler will generate just one instance of it and you can just include the header to point to that address

---
In reply to: [281726248](https://github.com/Azure/azure-iot-sdk-c/pull/984#discussion_r281726248) [](ancestors = 281726248,281403483,280640096)

**Comment by massand** [comment]
I have reached out to our service team to help us enable this feature in our Canary hub. Please add the e2e tests here or create a task to track this work. Leaving this active until either of that is done. Thanks.

---
In reply to: [490160932](https://github.com/Azure/azure-iot-sdk-c/pull/984#issuecomment-490160932) [](ancestors = 490160932,489824105,488879595)

File: provisioning_client/tests/prov_transport_http_client_ut/prov_transport_http_client_ut.c (hunk 5 above)
Comment (massand) [reply provisioning_client/tests/prov_transport_http_client_ut/prov_transport_http_client_ut.c]: I am only seeing the json always being returned is TEST_JSON_CONTENT (which doesn't have any ek or srk values in it). I am pretty sure we aren't exercising the construct_json_data function and testing cleanup on line 258 of prov_transport_http_client.c..

---
In reply to: [281715420](https://github.com/Azure/azure-iot-sdk-c/pull/984#discussion_r281715420) [](ancestors = 281715420,281249536)

File: provisioning_client/src/prov_device_ll_client.c [hunk 8]
```diff
@@ -1121,3 +1336,47 @@ const char* Prov_Device_LL_GetVersionString(void)
...
+        if (mallocAndStrcpy_s(&temp_data, jsonDataField) != 0)
+        {
+            LogError("Failure setting custom provisioning data");
+            result = PROV_DEVICE_RESULT_ERROR;
+        }
+        else
+        {
+            if (handle->custom_request_data != NULL)
+            {
+                free(handle->custom_request_data);
```
Comment (ewertons) [inline provisioning_client/src/prov_device_ll_client.c]: Nice

File: provisioning_client/src/prov_transport_utils.c [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+
+// Retry-After: Fri, 31 Dec 1999 23 : 59 : 59 GMT
+// Retry-After : 120
+uint32_t parse_retry_after_value(const char* retry_after)
+{
+    uint32_t result = PROV_GET_THROTTLE_TIME;
+    if (retry_after != NULL)
+    {
+        // Is the retry after a number
+        if (retry_after[0] >= 0x30 || retry_after[0] <= 0x39)
```
Comment (ewertons) [inline provisioning_client/src/prov_transport_utils.c]: Could you move these hex values into consts or defines upwards to make it more explicit what they mean?

File: provisioning_client/tests/prov_transport_amqp_client_ut/prov_transport_amqp_client_ut.c [hunk 10]
```diff
@@ -402,10 +404,10 @@ BEGIN_TEST_SUITE(prov_transport_amqp_client_ut)
     TEST_FUNCTION(prov_transport_amqp_register_device_succeed)
     {
         //arrange
-        STRICT_EXPECTED_CALL(prov_transport_common_amqp_register_device(TEST_DPS_HANDLE, on_transport_json_parse, NULL));
+        STRICT_EXPECTED_CALL(prov_transport_common_amqp_register_device(TEST_DPS_HANDLE, on_transport_json_parse, on_transport_create_json_payload, NULL));
 
         //act
-        int result = prov_amqp_transport_register_device(TEST_DPS_HANDLE, on_transport_json_parse, NULL);
+        int result = prov_amqp_transport_register_device(TEST_DPS_HANDLE, on_transport_json_parse, on_transport_create_json_payload, NULL);
```
Comment (ewertons) [inline provisioning_client/tests/prov_transport_amqp_client_ut/prov_transport_amqp_client_ut.c]: Missing negative test when callback is NULL.

File: provisioning_client/tests/prov_transport_amqp_ws_client_ut/prov_transport_amqp_ws_client_ut.c [hunk 11]
```diff
@@ -416,10 +418,10 @@ BEGIN_TEST_SUITE(prov_transport_amqp_ws_client_ut)
     TEST_FUNCTION(prov_transport_amqp_register_device_succeed)
     {
         //arrange
-        STRICT_EXPECTED_CALL(prov_transport_common_amqp_register_device(TEST_DPS_HANDLE, on_transport_json_parse, NULL));
+        STRICT_EXPECTED_CALL(prov_transport_common_amqp_register_device(TEST_DPS_HANDLE, on_transport_json_parse, on_transport_create_json_payload, NULL));
 
         //act
-        int result = prov_amqp_ws_transport_register_device(TEST_DPS_HANDLE, on_transport_json_parse, NULL);
+        int result = prov_amqp_ws_transport_register_device(TEST_DPS_HANDLE, on_transport_json_parse, on_transport_create_json_payload, NULL);
```
Comment (ewertons) [inline provisioning_client/tests/prov_transport_amqp_ws_client_ut/prov_transport_amqp_ws_client_ut.c]: Looks like it is missing a negative test when create_json_payload callback is NULL

File: provisioning_client/tests/prov_transport_http_client_ut/prov_transport_http_client_ut.c [hunk 12]
```diff
@@ -1349,7 +1369,7 @@ BEGIN_TEST_SUITE(prov_transport_http_client_ut)
...
         (void)prov_dev_http_transport_open(handle, TEST_REGISTRATION_ID_VALUE, TEST_BUFFER_VALUE, TEST_BUFFER_VALUE, on_transport_register_data_cb, NULL, on_transport_status_cb, NULL, on_transport_challenge_callback, NULL);
-        (void)prov_dev_http_transport_register_device(handle, on_transport_json_parse, NULL);
+        (void)prov_dev_http_transport_register_device(handle, on_transport_json_parse, on_transport_create_json_payload, NULL);
```
Comment (ewertons) [inline provisioning_client/tests/prov_transport_http_client_ut/prov_transport_http_client_ut.c]: Also missing negative test?

File: provisioning_client/tests/prov_transport_mqtt_ws_client_ut/prov_transport_mqtt_ws_client_ut.c (hunk 7 above)
Comment (ewertons) [inline provisioning_client/tests/prov_transport_mqtt_ws_client_ut/prov_transport_mqtt_ws_client_ut.c]: Missing negative test when the new callback is NULL.

File: provisioning_client/src/prov_device_ll_client.c [hunk 13]
```diff
@@ -251,13 +261,27 @@ static int retrieve_json_number(JSON_Object* json_object, const char* field_name
...
+}
+
+static char* retrieve_json_string(JSON_Object* json_object, const char* field_name, bool is_required)
 {
     char* result;
     JSON_Value* json_field;
     if ((json_field = json_object_get_value(json_object, field_name)) == NULL)
     {
-        LogError("failure retrieving json operation id");
+        if (is_required)
```
[the remark this answers is not in the document]
Comment (jebrando) [reply provisioning_client/src/prov_device_ll_client.c]: valid point, but I was to differentiate error so I want it to log.

---
In reply to: [283054389](https://github.com/Azure/azure-iot-sdk-c/pull/984#discussion_r283054389) [](ancestors = 283054389)

File: provisioning_client/tests/prov_transport_http_client_ut/prov_transport_http_client_ut.c (hunk 5 above)
Comment (jebrando) [reply provisioning_client/tests/prov_transport_http_client_ut/prov_transport_http_client_ut.c]: All the parsing happens in the prov_client_ll.c file through the callback.  So it doesn't matter what we return here, it just get passed to the callback.

---
In reply to: [282243402](https://github.com/Azure/azure-iot-sdk-c/pull/984#discussion_r282243402) [](ancestors = 282243402,281715420,281249536)
